### PR TITLE
Feature/dispute proof

### PR DESCRIPTION
**Pull request by sdaveas**
# tl;dr
Submit proof does not need to be verified on-chain anymore

# Functionality
There are two reasons a proof should be rejected: 
1. existing proof is not valid
2. existing proof is valid, but there is a larger, honest proof that does not include the claimed predicate

Previously, the first functionality was included in the submit-phase. Submit- and dispute-phases can be decoupled in order to have a more gas-efficient and more clear solution. When a proof is submitted, another party has the option to dispute its validity. During dispute-phase, the existing proof is validated applying a merkle-proof validation at each block. This removes the proof validation from the submit-proof phase.

# Further optimizations
Further optimizations can be made. The dispute operation can be restricted to a slice of the existing proof. The rationale is that once an invalid proof has been observed by a party, they only need to prove that the existing chain is flawed at a certain index of the proof which is an equivalent to the statement that the entire proof is invalid. The party that chooses to dispute an existing proof indicates the index in which the existing proof is invalid, resulting to constant time verification of the existing proof. Of course, one can chose to dispute the full range of the existing proof, which is no longer a constant operation.

### CODE REVIEW

File: experiment/contractNipopow.sol [hunk 1]
```diff
@@ -189,11 +193,16 @@ contract Crosschain {
                 reversedSiblings[j] = siblings[ptr + j];
             ptr += branchLength;
 
+            // We only care to verify within our range
+            if (i < startValidateIndex || endValidateIndex < i) {
+                continue;
+            }
```
Comment (gtklocker) [inline experiment/contractNipopow.sol]: This can be directly incorporated in the for loop

**Review by gtklocker — CHANGES_REQUESTED** [review]

File: experiment/contractNipopow.sol (hunk 1 above)
Comment (sdaveas) [reply experiment/contractNipopow.sol]: We have to iterate through siblings starting from `i=1`and update the `ptr`, because siblings array is not aligned with the headers array.

Comment (sdaveas) [reply experiment/contractNipopow.sol]: But this is ugly.
I can keep `validateInterlink` as it was before and have a different function to find the `siblings` offset and then call directly `validateMerkle`.
